sbi/utils/restriction_estimator.py

```diff
@@ -556,94 +513,66 @@ def density_thresholder(theta: Tensor) -> Tensor:
-            tf_min_val = torch.max(1.0 - clf_probs)
-            tf_estimate = tf_min_val + tf_min_val / num_valid
-            classifier_thr = 1.0 - tf_estimate
-        else:
-            raise NameError(f"`safety_margin` {safety_margin} not supported.")
-    else:
         assert (
-            allowed_false_negatives is not None
-        ), "`allowed_false_negatives` must be set."
-        quantile_index = floor(num_valid * allowed_false_negatives)
-        classifier_thr, _ = torch.kthvalue(clf_probs, quantile_index + 1)
-
-    classifier_thr = classifier_thr.detach()
-
-    def classifier_thresholder(theta: Tensor) -> Tensor:
-        pred = F.softmax(classifier.forward(theta), dim=1)[:, 1]
-        if reweigh_factor is None:
-            threshold = classifier_thr
+            allowed_false_negatives is None or reweigh_factor is None
+        ), """Both the `allowed_false_negatives` and the `reweigh_factor` are set. You
+            can only set one of them."""
+
+        valid_val_theta = validation_theta[validation_label.bool()]
+        num_valid = valid_val_theta.shape[0]
+        clf_probs = F.softmax(classifier.forward(valid_val_theta), dim=1)[:, 1]
+
+        if allowed_false_negatives == 0.0:
+            if safety_margin is None:
+                self._classifier_thr = torch.min(clf_probs)
+            elif isinstance(safety_margin, float):
+                self._classifier_thr = torch.min(clf_probs) - safety_margin
+            elif safety_margin == "frequentist":
+                # We seek the minimum classifier output, not the maximum, as it usually
+                # is in the `German Tank Problem`. Hence, we transform the outputs with
+                # (1-output), apply the estimator, and then transform back.
+                tf_min_val = torch.max(1.0 - clf_probs)
+                tf_estimate = tf_min_val + tf_min_val / num_valid
+                self._classifier_thr = 1.0 - tf_estimate
+            else:
+                raise NameError(f"`safety_margin` {safety_margin} not supported.")
+        else:
+            assert (
+                allowed_false_negatives is not None
+            ), "`allowed_false_negatives` must be set."
+            quantile_index = floor(num_valid * allowed_false_negatives)
+            self._classifier_thr, _ = torch.kthvalue(clf_probs, quantile_index + 1)
+
+    def __call__(self, theta):
+        pred = F.softmax(self._classifier.forward(theta), dim=1)[:, 1]
+        if self._reweigh_factor is None:
+            threshold = self._classifier_thr
             predictions = pred > threshold
         else:
-            probs_invalid = pred * reweigh_factor
-            probs_valid = (1 - pred) * (1 - reweigh_factor)
+            probs_invalid = pred * self._reweigh_factor
+            probs_valid = (1 - pred) * (1 - self._reweigh_factor)
             predictions = probs_valid > probs_invalid
         return predictions.bool()
 
-    if print_fp_rate:
```

Comment:
why did this get removed? `self._print_fp_rate` seems to be unused now.

Reply:
Deleting if print_fp_rate: was a mistake on my end. I have now included `print_false_positive_rate` as a method of `AcceptRejectFunction`.
Now, `if self._print_fp_rate`, the function is called. The refactoring is a bit awkward on this one, technically the method does not need any input parameters other than self but I opted to keep the input parameters so they remain explicit during the call and for consistency with the previous implementation. I also checked that setting it to true actually prints the fp rate. It’s also worth nothing that there seems to be no good way right now for a user to print the fp rate but since that’s unrelated to pickling I won’t go into detail here.
Fixing commit is below.